### PR TITLE
Implement hash on struct/list

### PR DESCRIPTION
**Pull request by acquamarin**
This PR implements hash operation on struct and list.
Note: hash on list of struct is not supported

### CODE REVIEW

File: src/include/function/hash/hash_functions.h [hunk 1]
```diff
@@ -54,99 +56,178 @@ struct CombineHash {
...
+    const common::list_entry_t& key, common::hash_t& result, common::ValueVector* keyVector) {
+    auto dataVector = common::ListVector::getDataVector(keyVector);
+    result = NULL_HASH;
+    common::hash_t tmpResult;
+    for (auto i = 0u; i < key.size; i++) {
+        auto pos = key.offset + i;
+        if (dataVector->isNull(pos)) {
+            result = combineHashScalar(result, NULL_HASH);
+        } else {
+            switch (dataVector->dataType.getPhysicalType()) {
```
Comment (ray6080) [inline src/include/function/hash/hash_functions.h:172]: Check if you can use `TypeUtils::visit` to simplify the switch/case code here.

File: src/processor/operator/aggregate/aggregate_hash_table.cpp [hunk 2]
```diff
@@ -639,6 +639,52 @@ static bool compareEntry(common::ValueVector* vector, uint32_t vectorPos, const
...
+bool compareEntry<list_entry_t>(
+    common::ValueVector* vector, uint32_t vectorPos, const uint8_t* entry) {
+    auto dataVector = ListVector::getDataVector(vector);
+    auto listToCompare = vector->getValue<list_entry_t>(vectorPos);
+    auto listEntry = reinterpret_cast<const ku_list_t*>(entry);
+    auto entryNullBytes = reinterpret_cast<uint8_t*>(listEntry->overflowPtr);
+    auto entryValues = entryNullBytes + NullBuffer::getNumBytesForNullValues(listEntry->size);
+    auto rowLayoutSize = LogicalTypeUtils::getRowLayoutSize(dataVector->dataType);
+    compare_function_t compareFunc;
+    AggregateHashTable::getCompareEntryWithKeysFunc(dataVector->dataType, compareFunc);
```
Comment (andyfengHKU) [inline src/processor/operator/aggregate/aggregate_hash_table.cpp:652]: Let's discuss if we can make wrap compare function as a nested data structure so we don't need to resolve it at compile time.

File: src/include/function/hash/hash_functions.h (hunk 1 above)
Comment (andyfengHKU) [reply src/include/function/hash/hash_functions.h:172]: Plus let's have decent coverage on this code path.

File: src/function/vector_hash_functions.cpp [hunk 3]
```diff
@@ -62,19 +62,29 @@ void VectorHashFunction::computeHash(ValueVector* operand, ValueVector* result)
...
-        // LCOV_EXCL_START
+    } break;
+    case PhysicalTypeID::VAR_LIST: {
+        // TODO(Ziyi): We should pass in the selection state here, and do vectorized hash
+        // computation.
+        UnaryHashFunctionExecutor::execute<list_entry_t, hash_t>(*operand, *result);
+    } break;
     default: {
-        throw RuntimeException("Cannot hash data type " + operand->dataType.toString());
+        KU_UNREACHABLE;
```
Comment (andyfengHKU) [inline src/function/vector_hash_functions.cpp:86]: Let's keep the original exception here. We don't have a protection in binder about what we can hash and we cannot. So it's better to have a meaningful error message.

**Review by andyfengHKU — APPROVED** [review]

**Pull request merged**